source/features/select-notifications.tsx

```diff
@@ -0,0 +1,176 @@
+import React from 'dom-chef';
+import select from 'select-dom';
+import * as pageDetect from 'github-url-detection';
+import {
+	CheckCircleIcon,
+	CheckIcon,
+	DotFillIcon,
+	DotIcon,
+	GitMergeIcon,
+	GitPullRequestDraftIcon,
+	GitPullRequestIcon,
+	IssueOpenedIcon,
+	XCircleIcon
+} from '@primer/octicons-react';
+
+import features from '.';
+
+const filters = {
+	'Pull requests': ':is(.octicon-git-pull-request, .octicon-git-pull-request-closed, .octicon-git-pull-request-draft, .octicon-git-merge)',
+	Issues: ':is(.octicon-issue-opened, .octicon-issue-closed)',
+	Open: ':is(.octicon-issue-opened, .octicon-git-pull-request)',
+	Closed: ':is(.octicon-issue-closed, .octicon-git-pull-request-closed)',
+	Draft: '.octicon-git-pull-request-draft',
+	Merged: '.octicon-git-merge',
+	Read: '.notification-read',
+	Unread: '.notification-unread'
+};
+
+type Filter = keyof typeof filters;
+type Category = 'Type' | 'Status' | 'Read';
+
+function resetFilters({target}: Event): void {
+	for (const checkbox of select.all('input[type="checkbox"]:checked', target as Element)) {
+		checkbox.checked = false;
+		checkbox.parentElement!.parentElement!.setAttribute('aria-checked', 'false');
+	}
+}
+
+function handleSelection({target}: Event): void {
+	const selectedFilters = select.all('[aria-checked="true"]', target as Element);
+	const selectAllCheckbox = select('input[type="checkbox"].js-notifications-mark-all-prompt')!;
+	// Reset the "Select all" checkbox
+	if (selectAllCheckbox.checked) {
+		selectAllCheckbox.dispatchEvent(new MouseEvent('click'));
+	}
+
+	if (selectedFilters.length > 0) {
+		const activeFilters: Record<Category, string[]> = {
+			Type: [],
+			Status: [],
+			Read: []
+		};
+		for (const selectedFilter of selectedFilters) {
+			activeFilters[selectedFilter.dataset.category as Category].push(filters[selectedFilter.dataset.filter as Filter]);
+		}
+
+		for (const notification of select.all('.notifications-list-item')) {
+			let isNotificationSelected = true;
+			for (const [category, categoryFilters] of Object.entries(activeFilters)) {
+				if (category === 'Read') {
+					if (categoryFilters.length === 1 && !notification.matches(categoryFilters[0])) {
+						isNotificationSelected = false;
+						break;
+					}
+
+					continue;
+				}
+
+				if (categoryFilters.length > 0 && !select.exists(categoryFilters, notification)) {
+					isNotificationSelected = false;
+					break;
+				}
+			}
+
+			// Make excluded notifications unselectable
+			if (!isNotificationSelected) {
+				select('.js-notification-bulk-action-check-item', notification)!.removeAttribute('data-check-all-item');
+			}
+		}
+
+		// If at least one notification is selectable, trigger the "Select all" checkbox
+		if (select.exists('.js-notification-bulk-action-check-item[data-check-all-item]')) {
+			selectAllCheckbox.dispatchEvent(new MouseEvent('click'));
+		}
```

Comment:
I used this roundabout way of selecting the notifications because:
1. We can't just set the `checked` attribute of their checkboxes (it doesn't update the "Select all" count).
2. Dispatching a `MouseEvent` to every selected notification causes some noticeable lag, when a lot of notifications are selected at once.

Reply:
Would it be possible to set the `checked` property and then trigger the count once at the end?

Reply:
That was my first thought, sadly it doesn't work (every checkbox ends up selected regardless).